Software update banner

wmanley opened the pull request:
New look software update banner, refactoring HTTP polling.

I wasn't sure how to write automated tests for this, please point me in the right direction.

Screenshot:

![screenshot from 2015-01-08 16 06 31](https://cloud.githubusercontent.com/assets/494659/5665502/c71e9732-9750-11e4-8fba-55f0a1ea4434.png)

drothlis commented:
I had envisioned the notification going *below* the navbar (completely outside the navbar div).

drothlis commented on app/app.js:
```diff
@@ -147,32 +144,47 @@ app.controller('LiveController', function($http, $scope) {
+      "/software-update/update-available", 60*1000, $http, $interval);
 });
 
+function http_poll(endpoint, interval, $http, $interval)
```
So far I've kept the Controllers at the top of the file, with helper functions out of the way at the bottom.

drothlis commented on app/index.html:
```diff
@@ -14,7 +14,22 @@
+<div class="navbar navbar-default navbar-static-top" role="navigation" bs-navbar>
+    <div class="container-fluid alert alert-success" ng-controller="VersionController"
+         ng-show="software_update_available.value"
+         class="alert alert-success" role="alert" style="padding-top: 0px; padding-bottom:0px; margin-top: 0px; margin-bottom: 0px; border-radius: 0px">
```
Keep the css out of the html file; in app.css instead.

drothlis commented on app/index.html:
```diff
@@ -14,7 +14,22 @@
+        <a href="http://stb-tester.com/stb-tester-one/rev2015.1/release-notes.html"
+           class="alert-link">
+          release notes</a> for more information.
+          <button style="margin-left: 3em" type='submit' class='btn btn-success'>
```
Same here.

drothlis commented on app/index.html:
```diff
@@ -14,7 +14,22 @@
 
 <!-- Navbar ----------------------------------------------------------------->
-<div class="navbar navbar-default navbar-fixed-top" role="navigation" bs-navbar>
+<div class="navbar navbar-default navbar-static-top" role="navigation" bs-navbar>
```
I think this (navbar not fixed) is a good idea even if we put the notification area outside & below the navbar div. So that it works with narrow windows (phones?) as you pointed out. It also simplifies things (less css, less calculations to do when working out desired scroll positions in JavaScript which we'll need to do when highlighting the current line of the script being run, etc.)

drothlis commented:
> I wasn't sure how to write automated tests for this, please point me in the right direction.

Integration / UI tests are in `e2e-tests/scenarios.py` and are run with `make check-all`. See `README.md` for the dependencies you have to install.

I'm not sure how we want to test this -- the test could ssh to the appliance & touch the `update-available` file & check that the notification appears. However it would be a very long-running test because the web-ui only checks for the file once a minute. So it hardly seems worth testing. It'll be worth testing if we actually test the complete update process, from publishing a new release to checking the notification, clicking the "install" button, and checking that the new version is installed.

Make sure you rebase onto master because the e2e-tests have changed significantly since your branch point.

wmanley commented:
> I had envisioned the notification going below the navbar (completely outside the navbar div).

Yeah, me too. I'd done it like that because it was the easiest to do.  I've fixed that now:

![screenshot from 2015-01-08 20 42 14](https://cloud.githubusercontent.com/assets/494659/5670373/29de7088-9777-11e4-8cc0-b7a3bc6830ee.png)

and it looks much better.

I've fixed the CSS and helper function location too.

I agree the software update test would be more appropriate if it were a true end-to-end test.

drothlis commented on app/index.html:
```diff
@@ -40,6 +40,23 @@
 </div>
 
+<div class="container-fluid alert alert-success top-alert"
+     ng-controller="VersionController"
```
If I understand the Angular docs correctly, this means that we'll have *two* instances of `VersionController`, each polling once per minute and updating its own copy of `$scope.software_update_available`. (The other instance is from this in the navbar: `<p ng-controller="VersionController">{{version}}</p>`.)

It shouldn't actually *matter* and we can fix it by hard-coding the version into the html at build time, and we don't need to fix it before merging this pull request.

drothlis commented:
Looking good.
